astropy/config/configuration.py

```diff
@@ -460,6 +460,7 @@ def section_name(section):
                     f" deprecated. Use '{self.name}'"
                     f" {section_name(new_module)} instead.",
                     AstropyDeprecationWarning,
+                    stacklevel=2,
```

Comment:
What? Why?

Reply:
I am equally puzzled

Reply:
The idea is to show more useful information to the user, with `stacklevel=2` it will show the caller line instead of the `warnings.warn()` line:
```
In [1]: import warnings

In [2]: def foo():
   ...:     warnings.warn('be careful')

In [3]: foo()
<ipython-input-2-1a647199d0e9>:2: UserWarning: be careful
  warnings.warn('be careful')

In [4]: def foo():
   ...:     warnings.warn('be careful', stacklevel=2)

In [5]: foo()
<ipython-input-5-c19b6d9633cf>:1: UserWarning: be careful
  foo()
```

However this is not so simple. First because warning are catched and printed by the logger, so this information is lost anyway. And even if someone disable this, the information printed with `stacklevel=2` may not be useful when the warning is emitted deep in the library:
```
In [2]: astropy.log.disable_warnings_logging()

In [3]: hdr = fits.Header()

In [4]: hdr['LONG KEYWORD'] = 'hello world'
/home/simon/dev/astropy/astropy/astropy/io/fits/card.py:199: VerifyWarning: Keyword name 'LONG KEYWORD' is greater than 8 characters or contains characters not allowed by the FITS standard; a HIERARCH card will be created.
  self.keyword = keyword

In [5]: hdr['KEYWORD'] = ('hello world', 'comment'*10)

In [6]: hdr
Out[6]: /home/simon/dev/astropy/astropy/astropy/io/fits/card.py:531: VerifyWarning: Card is too long, comment will be truncated.
  self._image = self._format_image()

HIERARCH LONG KEYWORD = 'hello world'                                           
KEYWORD = 'hello world'        / commentcommentcommentcommentcommentcommentcomme
```
`self.keyword = keyword` or `self._image = self._format_image()` don't really help.

Reply:
> stacklevel=2 it will show the caller line instead of the warnings.warn()

That was my intention, to provide slightly more information, which is hopefully more useful.

>  And even if someone disable this, the information printed with stacklevel=2 may not be useful when the warning is emitted deep in the library:

A good point. Maybe not all should be `stacklevel=2`. But definitely some should be. Ugh. That's more of a pain. This should probably be done on a module-by-module basis. I'll remove that commit for now.

Reply:
Removed.

Reply:
In general we probably shouldn't make this kind of change in a PR that is otherwise about style - best done in a dedicated PR